cookbook/controller/service.rst

```diff
@@ -232,6 +234,34 @@ inject *only* the exact service(s) that you need directly into the controller.
    are valid, exactly how you want to organize your reusable code is up to
    you.
 
+Base Controller Methods and their Service Replacements
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This table explains how to replace the convenience methods of the base
+controller.
+
++-----------------------------+------------------------------------+----------------------------------------------------------------+
+| Method                      | Service                            |  PHP Code                                                      |
++=============================+====================================+================================================================+
+| ``createForm``              | ``form.factory``                   | ``$formFactory->create($type, $data, $options)``               |
+| ``createFormBuilder``       | ``form.factory``                   | ``$formFactory->createBuilder('form', $data, $options)``       |
+| ``createNotFoundException`` | -                                  | ``throw new NotFoundHttpException($message, $previous);``      |
+| ``forward``                 | ``http_kernel``                    | ``$httpKernel->forward($controller, $path, $query)``           |
+| ``generateUrl``             | ``router``                         | ``$router->generate($route, $params, $absolute)``              |
+| ``getDoctrine``             | ``doctrine``                       |                                                                |
+| ``isGranted``               | ``security.authorization_checker`` | ``$authorizationChecker->isGranted($attributes, $object);``    |
+| ``redirect``                | -                                  | ``return new RedirectResponse($url, $status);``                |
+| ``render``                  | ``templating``                     | ``$templating->renderResponse($view, $parameters, $response)`` |
+| ``renderViev``              | ``templating``                     | ``$templating->render($view, $parameters)``                    |
++-----------------------------+------------------------------------+----------------------------------------------------------------+-
+
+.. tip::
+
+    ``getRequest`` has been deprecated. Instead, have an argument to your
+    controller action method called ``Request $request``. The order of the
+    parameters is not important, but the typehint must be provided.
```

Comment:
converted this to a separate tip. i think it does not hurt to mention this here again. the whole doc segment is for people afraid to look at the controller source code.